File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/livestream/state/UserState.java
package com.trtc.uikit.livekit.livestream.state;

import android.text.TextUtils;

import androidx.annotation.Nullable;

import com.tencent.cloud.tuikit.engine.room.TUIRoomDefine;
import com.trtc.tuikit.common.livedata.LiveData;

import java.util.LinkedHashSet;
import java.util.Objects;
import java.util.Set;

public class UserState {
    public UserInfo                          selfInfo               = new UserInfo();
    public LiveData<LinkedHashSet<UserInfo>> userList               = new LiveData<>(new LinkedHashSet<>());
    public LiveData<LinkedHashSet<String>>   hasAudioStreamUserList = new LiveData<>(new LinkedHashSet<>());
    public LiveData<LinkedHashSet<String>>   hasVideoStreamUserList = new LiveData<>(new LinkedHashSet<>());
    public LiveData<LinkedHashSet<String>>   speakingUserList       = new LiveData<>(new LinkedHashSet<>());
    public LiveData<LinkedHashSet<UserInfo>> myFollowingUserList    = new LiveData<>(new LinkedHashSet<>());
    public LiveData<UserInfo>                enterUserInfo          = new LiveData<>();

    public void reset() {
        userList.get().clear();
        hasAudioStreamUserList.get().clear();
        hasVideoStreamUserList.get().clear();
        speakingUserList.get().clear();
        myFollowingUserList.get().clear();
    }

    public void addUserList(Set<UserInfo> list) {
        if (list == null || list.isEmpty()) {
            return;
        }
        userList.addAll(list);
    }

    public void addUser(TUIRoomDefine.UserInfo userInfo) {
        if (userInfo == null) {
            return;
        }
        if (TextUtils.isEmpty(userInfo.userId)) {
            return;
        }
        UserInfo info = new UserInfo(userInfo.userId);
        enterUserInfo.set(info);
        userList.add(info);
    }

    public void removeUser(TUIRoomDefine.UserInfo userInfo) {
        if (userInfo == null) {
            return;
        }
        if (TextUtils.isEmpty(userInfo.userId)) {
            return;
        }
        userList.remove(new UserInfo(userInfo.userId));
    }

    public static class UserInfo {
        public String                       userId    = "";
        public LiveData<String>             name      = new LiveData<>("");
        public LiveData<String>             avatarUrl = new LiveData<>("");
        public LiveData<TUIRoomDefine.Role> role      = new LiveData<>(TUIRoomDefine.Role.GENERAL_USER);
        public LiveData<Long>               fansCount = new LiveData<>(0L);

        public UserInfo() {
        }

        public UserInfo(String userId) {
            this.userId = userId;
        }

        public UserInfo(TUIRoomDefine.UserInfo userInfo) {
            updateState(userInfo);
        }

        @Override
        public boolean equals(@Nullable Object obj) {
            if (obj instanceof UserInfo) {
                return this.userId.equals(((UserInfo) obj).userId);
            }
            return false;
        }

        @Override
        public int hashCode() {
            return Objects.hash(userId);
        }

        public void updateState(TUIRoomDefine.UserInfo userInfo) {
            this.userId = userInfo.userId;
            this.name.set(userInfo.userName);
            this.avatarUrl.set(userInfo.avatarUrl);
            this.role.set(userInfo.userRole);
        }
    }
}

File: Flutter/lib/state/operation/user_state.dart
import 'dart:collection';
import 'package:flutter/material.dart';
import 'package:rtc_room_engine/api/room/tui_room_define.dart';

class UserState {
  final UserInfo selfInfo = UserInfo();
  final ValueNotifier<LinkedHashSet<UserInfo>> userList = ValueNotifier(LinkedHashSet<UserInfo>());
  final ValueNotifier<LinkedHashSet<String>> hasAudioStreamUserList = ValueNotifier(LinkedHashSet<String>());
  final ValueNotifier<LinkedHashSet<String>> hasVideoStreamUserList = ValueNotifier(LinkedHashSet<String>());
  final ValueNotifier<LinkedHashSet<String>> hasScreenStreamUserList = ValueNotifier(LinkedHashSet<String>());
  final ValueNotifier<LinkedHashSet<String>> speakingUserList = ValueNotifier(LinkedHashSet<String>());
  final ValueNotifier<LinkedHashSet<UserInfo>> myFollowingUserList = ValueNotifier(LinkedHashSet<UserInfo>());

  void reset() {
    userList.value.clear();
    hasAudioStreamUserList.value.clear();
    hasVideoStreamUserList.value.clear();
    hasScreenStreamUserList.value.clear();
    speakingUserList.value.clear();
    myFollowingUserList.value.clear();
  }

  void addUserList(Set<UserInfo> list) {
    if (list.isEmpty) {
      return;
    }
    userList.value.addAll(list);
    final userListValue = LinkedHashSet<UserInfo>();
    userListValue.addAll(list);
    userList.value = userListValue;
  }

  void addUser(TUIUserInfo userInfo) {
    if (userInfo.userId.isEmpty) {
      return;
    }
    userList.value.add(UserInfo.fromTUIUserInfo(userInfo));
    final userListValue = LinkedHashSet<UserInfo>();
    userListValue.addAll(userList.value);
    userList.value = userListValue;
  }

  void removeUser(TUIUserInfo userInfo) {
    if (userInfo.userId.isEmpty) {
      return;
    }
    userList.value.removeWhere((element) => element.userId == userInfo.userId);
    final userListValue = LinkedHashSet<UserInfo>();
    userListValue.addAll(userList.value);
    userList.value = userListValue;
  }
}

class UserInfo {
  String userId = '';
  ValueNotifier<String?> name = ValueNotifier('');
  ValueNotifier<String?> avatarUrl = ValueNotifier('');
  ValueNotifier<TUIRole> role = ValueNotifier(TUIRole.generalUser);
  ValueNotifier<int> fansCount = ValueNotifier(0);

  UserInfo();

  UserInfo.formUserId(this.userId);

  UserInfo.fromTUIUserInfo(TUIUserInfo userInfo) {
    updateState(userInfo);
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is UserInfo && other.userId == userId;
  }

  @override
  int get hashCode => userId.hashCode;

  void updateState(TUIUserInfo userInfo) {
    userId = userInfo.userId;
    name.value = userInfo.userName;
    avatarUrl.value = userInfo.avatarUrl;
    role.value = userInfo.userRole;
  }
}
